go/vt/vtgate/scatter_conn.go

```diff
@@ -366,9 +366,15 @@ func (stc *ScatterConn) StreamExecuteMulti(
 
 // Commit commits the current transaction. There are no retries on this operation.
 func (stc *ScatterConn) Commit(context context.Context, session *SafeSession) (err error) {
+	if session == nil {
+		return fmt.Errorf("cannot commit: empty session")
+	}
 	if !session.InTransaction() {
 		return fmt.Errorf("cannot commit: not in transaction")
 	}
+	if len(session.ShardSessions) == 0 {
+		return fmt.Errorf("cannot commit: invalid session, always update session from RPC response")
```

Comment:
I discussed this with Liang. In this particular case, we don't know if the client wanted to explicitly commit an empty transaction, which is legitimate, or if it had a bug.
Also, as a bug, this is just a degenerate case of a partially updated session. The other cases are impossible to detect, and can lead to equally bad data loss.
So, I think we should just clearly document the contract, and put warning signs there about this pitfall.